### PR TITLE
Compound object support - 2nd try

### PR DESCRIPTION
**Pull request by patdunlavey**
JIRA Ticket: none

# Why:
* When viewing a compound object, the metadata and image for the pid currently being displayed usually do not make sense.
* This function can be called more than once if there are multiple objects being displayed on the page.
* schema.org itemscope div breaks element heights in bootstrap

# What does this Pull Request do?
* When viewing a compound object, the title is derived by concatenating the parent object title with the title of the current child object.
* When viewing a compound object, try to display the abstract for the currently displayed child object. If empty, then try to display the abstract of the parent.
* When viewing a compound object, use the image of the child being displayed.
* To prevent running multiple times, use a static variable to keep track whether it has run previously. If so, do nothing.
* Inject the schema.org itemscope div using ajax.

# How should this be tested?
* A handy open graph preview plugin is linked below.
* View a compound object. Do the metatags make sense for the object currently being displayed? 
* View another child of the compound object. Do the metatags make sense for this child?
* Regression: View other kinds of objects, do the metatags make sense for those objects?

# Reference Links:
* [A handy Chrome plugin for previewing](https://chrome.google.com/webstore/detail/open-graph-preview/ehaigphokkgebnmdiicabhjhddkaekgh?hl=en)
* https://github.com/Islandora-Labs/islandora_social_metatags
* https://github.com/Islandora-Labs/islandora_social_metatags/pull/8

Interested parties
@bondjimbond

### CODE REVIEW

**Comment by bondjimbond** [comment]
Thanks, @patdunlavey! Still ingesting a test object or two to try this out, but at first glance, it looks like this *only* works on compound objects. In the source code for a Citation object, none of the meta tags provided by the module appear. On a Compound object, they appear no problem.

**Comment by bondjimbond** [comment]
Update: Large Image works, so do others. Looks like the problem was with one particular object, not Citations in general. Will investigate further. Looks pretty good so far.

**Comment by bondjimbond** [comment]
Regarding the concatenated title piece: I can see the use case for this, but I can also see situations where someone might prefer the titles to stand on their own. Can we make this a configurable setting?

**Comment by patdunlavey** [comment]
Yes, that seems reasonable. I'll work on that.

**Comment by patdunlavey** [comment]
I made a bunch of changes. Configuration options have been added for how to handle titles and abstracts for compound objects. The logic for compound objects has been made a little more generalized so that one could, in theory, extend this idea for other kinds of hierarchical object relationships.
It occurs to me that it would probably make more sense to provide these enhancements through hooks, but I'm not sure when I would be able to get to that.

**Comment by bondjimbond** [comment]
Some weirdness happens when your Compound Object (parent) has an abstract, but so does its children. If you view the source of the parent Compound Object, you get in your Description tags:

> Parent Object Abstract :: First Child Abstract

e.g. `<meta property="og:description" content="This is the abstract :: This is the first child abstract." />`

**Comment by patdunlavey** [comment]
The configuration page should have settings for both the title and abstract.

**Comment by bondjimbond** [comment]
Ah, I missed that bit. Thanks!

If you select "child", then on the parent object, the first child object's description shows. If you select "parent", then the parent object's description shows on all objects including children.

We also need a way to display just the abstract of the object you're looking at (i.e. parent when looking at the parent, child when looking at a child).

File: includes/admin.form.inc [hunk 1]
```diff
@@ -66,5 +66,40 @@ function islandora_social_metatags_admin_form($form, $form_settings) {
...
+    '#options' => array(
+      'child' => t('Child'),
+      'parent' => t('Parent'),
+      'concatenate' => t('Both'),
+    ),
+    '#default_value' => variable_get('islandora_social_metatags_compound_abstract', 'concatenate'),
+  );
+  $form['islandora_social_metatags_compound_abstract_delimiter'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Compound Object Title Concatenation Delimiter'),
```
Comment (bondjimbond) [inline includes/admin.form.inc]: Should be "abstract concatenation delimiter"

**Comment by patdunlavey** [comment]
In normal compound object display, you are always viewing a child, whether you are on the compound object pid or that of one of the children. That child is displayed in the context of the compound object - it's parent.
So it makes sense to me that if the preference is 'parent', then you see the compound object's title/abstract; when you select 'child' you see the currently displayed child's title/abstract; and if you select 'concatenate' (or 'both'), you see both. 
Does that not make sense?

**Comment by bondjimbond** [comment]
Hmm, this is true. I'm thinking of cases where the compound and the children all have their own unique abstracts - which is possible, but perhaps unlikely/undesired. That should be OK then.

File: includes/admin.form.inc [hunk 2]
```diff
@@ -66,5 +66,40 @@ function islandora_social_metatags_admin_form($form, $form_settings) {
     '#description' => t('If the above Image datastreams do not exist for a given object, use a basic default. Use a relative path.'),
     '#default_value' => variable_get('islandora_social_metatags_noimage', '/sites/all/modules/islandora/images/folder.png'),
   );
+  $form['islandora_social_metatags_compound_title'] = array(
+    '#type' => 'select',
+    '#title' => t('Compound Object Title Preference'),
+    '#description' => t('Do you prefer the child title, parent title, or both, concatenated?.'),
```
Comment (bondjimbond) [inline includes/admin.form.inc]: Would reword slightly for clarity. Something to the effect of: "On all child objects, use either the Parent's title, the Child's title, or concatenate both. If choosing Child, the first child object's title will be used on the parent object."

**Comment by patdunlavey** [comment]
I suppose we could have another option: "Currently viewed object"? Description: "When viewing the pid of the parent compound object, show the title/abstract for the parent. When viewing the pid of one of a compound object's children, show the title/abstract of that child."
But it's already pretty complicated, and from the user's standpoint, there is no discernible difference between viewing the parent object and viewing the first child.

**Comment by patdunlavey** [comment]
Bit more wordsmithing done.

**Comment by bondjimbond** [comment]
Cool. This looks all right to me, then; I'll merge soon. This was a big modification -- thanks!

**Comment by patdunlavey** [comment]
Thanks Brandon!
